src/qibo/transpiler/optimizer.py

```diff
@@ -27,7 +27,10 @@ def __call__(self, circuit: Circuit) -> Circuit:
             )
         if logical_qubits == physical_qubits:
             return circuit
-        new_circuit = Circuit(physical_qubits)
+        new_wire_names = circuit.wire_names.copy() + [
+            name for name in self.connectivity.nodes if name not in circuit.wire_names
+        ]
+        new_circuit = Circuit(nqubits=physical_qubits, wire_names=new_wire_names)
```

Comment:
This could be the inspiration for a non-breaking change.

We could make `nqubits: Optional[int] = None`, and if the `wire_names` is specified, just compute `nqubits=len(wire_names)`.

In principle, we could even interpret the first argument in different ways according to the type, allowing the
```py
Circuit(["A", "D", "B", "C"])
```
syntax.

(There is even dedicated support in standard library for that, as [`@singledispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch), though we should check that is available also for methods, and in particular for `__init__` - though the second one is more likely if the first is true, since `__init__` is just a normal method, invoked by the class machinery, while the actual constructor is `__new__`)

Reply:
https://stackoverflow.com/a/34484742

Reply:
Actually, it's working:
```py
[ins] In [1]: from functools import singledispatchmethod
[nav] In [2]: class Ciao:
         ...:     @singledispatchmethod
         ...:     def __init__(self, arg):
         ...:         raise NotImplementedError
         ...:
         ...:     @__init__.register
         ...:     def _(self, arg: int):
         ...:         print(-arg)
         ...:
         ...:     @__init__.register
         ...:     def _(self, arg: bool):
         ...:         print(not arg)
[ins] In [3]: Ciao(30)
-30
Out[3]: <__main__.Ciao at 0x7ff57f710510>
[ins] In [4]: Ciao(False)
True
Out[4]: <__main__.Ciao at 0x7ff58418a3d0>
```
https://docs.python.org/3/library/functools.html#functools.singledispatchmethod

Reply:
> note that the dispatch happens on the type of the first non-self or non-cls argument:

I tried using @singledispatchmethod, but it requires the specific first argument to find dispatch, which conflicts with initializing circuit using keyword args. We need to add preprocessing before every ``circuit = Circuit(**init_kwargs)``.

The code was:

```py
class Circuit:
...
    @singledispatchmethod
    def __init__(
        self,
        nqubits: Optional[int],
        wire_names: Optional[list],
        accelerators = None,
        density_matrix: bool = False,
    ):
        NotImplementedError("This method should not be called directly.")
                            
    @__init__.register
    def _(self, nqubits: int, wire_names=None, accelerators=None, density_matrix=False):
        self._init(nqubits, wire_names, accelerators, density_matrix)
    
    @__init__.register
    def _(self, wire_names: list, nqubits=None, accelerators=None, density_matrix=False):
        self._init(len(wire_names), wire_names, accelerators, density_matrix)

    def _init(self, nqubits, wire_names, accelerators, density_matrix):
        if nqubits < 1:
            raise_error(
                ValueError,
                f"Number of qubits must be positive but is {nqubits}.",
            )
        ...
```

Reply:
Yes, the idea was indeed to have `nqubits: int`, and `nqubits: list` and interpret the `nqubits: list` as `wire_names` (or choose a more agnostic name to span the two of them).

In any case, the idea is to treat them as mutually exclusive arguments, since one can be inferred from the other. `@singledispatchmethod` was just an option, but you may also do it manually, applying a `None` default and checking that one of the two *has* to be `None`, and it will be inferred from the other.

> which conflicts with initializing circuit using keyword args.

In any case, this was exactly to allow not using only keyword args, but to allow passing both `nqubits` and `wire_names` as positional (since they are mutually alternative).
You can just keep use the signature:
```py
def __init__(self, nqubits: Optional[int, list] = None, ..., wire_names: Optional[list] = None):
    ...
```
and just parse the options by yourself:
- exactly one has to be different from `None`
- the other one is inferred
- if `nqubits` is a `list`, is interpreted as `wire_names`

The only pain point is that this will allow you to also do:
```py
Circuit(nqubits=["A", "B", "C", "D"])
```
which is a bit confusing, and it would have been ruled out by `single_dispatch` (otherwise you can not easily distinguish if an argument has been passed as positional or keyword).

Reply:
[93ac772](https://github.com/qiboteam/qibo/pull/1500/commits/93ac772c579f932e9d70fe7c79f4862b81da829f)

I think `qubits_spec` or `qubit` could work as agnostic names. Let me know if you have other names.

Reply:
I think it is better to handle in another PR, if needed.